### ISSUE TITLE
Any way to specify custom animations?

### ISSUE THREAD

**Issue by Nickersoft** [react-native-community/react-native-modal#70]
Hey guys, great library! I'm currently calling `initializeRegistryWithDefinitions()` on Animatable with my own animations, and tried using them in the "animationIn" and "animationOut" props of this modal. While it seems to work fine on iOS, I get the following error when running it on my Android device:

```
09-19 00:39:45.111  6186  6216 E ReactNativeJS: undefined is not a function (evaluating '_this.contentRef[_this.props.animationIn](_this.props.animationInTiming)')
```

Which basically suggests that my animation isn't initialized by the time the modal appears, despite `initializeRegistryWithDefinitions()` already being called. Any reason there would be such a behavioral difference between platforms?

**Comment by mmazzarolo** [comment]
Hey @Nickersoft, thank you for opening the issue.  
Since we don't have any platform check inside the modal, my first guess is that the issue might be somewhat related to react-native-animatable.  
Maybe your `initializeRegistryWithDefinitions()` call conflicts with our internal call [here](https://github.com/react-native-community/react-native-modal/blob/master/src/index.js#L10)?  🤔

**Comment by Nickersoft** [comment]
@mmazzarolo Yeah, I noticed that.. and actually, after running my code on Android then switching back to my iOS emulator, I was getting the same error there too. It almost seems like the only reason it was working before was a race condition of some sort (maybe my animations were the most recent to register?)

I was actually giving thought to adding custom animation support myself... however, I figured I'd start a discussion on what the API for that would look like. Passing in your own animation registry seems messy, plus I'm not even sure what you'd call that prop. It would be nice if you could pass in either the string of an existing animation, or an animation object like you would find in an `initializeRegistryWithDefinitions()` call. That said, this library's code already initializes the registry with custom animations so I'm not even sure what kind of refactoring would be required to make that happen. Either way, it's an important functionality for my app to have, so I'm determined to try to find a solution.

What are your thoughts on this?

**Comment by mmazzarolo** [comment]
@Nickersoft glad you're interested in sending a PR!  

I'll admit I'm not an expert of react-native-animatable, but I'll try to give you as much supported as I can. 
First of all, the call to `initializeRegistryWithDefinitions()` was added in PR #17 to fix a breaking change of the sliding animation in a react-native-animatable update.  

Regarding the `prop` to customize the animation: I'm not too keen on adding too many customization options on the component. 
Also, I still haven't put enough resources in learning how `initializeRegistryWithDefinitions()`, but the more I think about it, the more I think it should stay outside of the modal definition (since it initializes the registry globally, am I right?).  
Maybe showing in the modal README.md how to add a properly working slide animation pointing it to react-native-animatable docs might be enough?  

Still, in the meanwhile, considering that we already use the `initializeRegistryWithDefinitions()`, I guess it won't hurt adding a way to customize it too, so go for it. A prop name like `animatableDefinitions` should be good, what do you think?  

You would also need to move the initialization somewhere in the modal lifecycle, making sure that it is called with the current default definitions if `animatableDefinitions` is not provided and it is backward compatible 👍

**Comment by mmazzarolo** [comment]
Please note that if you're forking from master, there is still PR #65 that has not been published on npm (it should work thought, since it has been tested)

**Comment by Nickersoft** [comment]
Hey @mmazzarolo, so I poked around the code and was able to make it so now the `animationIn` and `animationOut` can accept either a string (which uses an animation from the default registry) or an animatable registry object in the following form (as you would when registering it globally):

```javascript
{
    from: {},
    to: {}
}
```

This approach leverages the `createAnimation()` and `registerAnimation()` methods of react-native-animatable, and seems to work for my purposes. I can open a PR, but first.. what are your thoughts on this? This way, people can export their animation definitions as objects and using them would be as simple as:

```jsx
import {myAnimations} from './helpers';
// ...
<Modal animationIn={myAnimations.fadeIn}>
```

I considered the `animatableDefinitions` prop, but thought it might be pointless for users to pass in their entire animation registry if they are only using two of the animations. Plus, this way it doesn't *require* people to use a registry, if they aren't using react-native-animatable in their app already.

**Comment by Nickersoft** [comment]
Oh, and you can check out my fix [here](https://github.com/Nickersoft/react-native-modal/commit/4140a807162b44809839c6065b689d8caed3de15).

**Comment by mmazzarolo** [comment]
That's neat, I like this approach even more than the one I suggested.  
Go with the PR 👌

**Comment by mmazzarolo** [comment]
One more think, could you also update the README too?  
Adding the prop to the list and pointing somewhere to the react-native-animatable docs (on creating custom animation) could be enough.

**Comment by Nickersoft** [comment]
Sure thing. I'll get that done and post a link to the PR when I'm done. One question though... I noticed there is some type information in `index.d.ts`. It's obviously TypeScript (which I've never used) but to me looks a lot like Facebook Flow type checking (which I have used). Is the syntax the same? If so, I can go and update that too, as I don't want any type errors being thrown. I already updated the PropTypes definition.

**Comment by Nickersoft** [comment]
All set: https://github.com/react-native-community/react-native-modal/pull/72

**Comment by mmazzarolo** [comment]
Yes, it's typescript typedefinition, but your change shouldn't break anything even without updating it.  
Maybe @iRoachie ( 👼 ) can update it once the PR gets merged?  
  
I also noticed one thing that can be improved in my opinion.  
Since your declaring the animation info as a component internal field (`this.animationOut = animationOut`) you won't be able to update it at runtime.  
I know that it might be something that doesn't interest you directly, but if you could handle this case too it would be awesome (I could give it a shot too otherwise).  
The easiest way to to it in my opinion is to move the logic you put in the constructor in a class method (something like `_buildAnimations = (props: Props => {}`) and call this method both in the `constructor` (passing the `props` as a parameter) and in `componentWillReceiveProps` (passing `nextProps` as a parameter).  
What do you think?

**Comment by iRoachie** [comment]
Sure thing @mmazzarolo

**Comment by Nickersoft** [comment]
Consider it done @mmazzarolo :) The PR is now updated.

**Comment by mmazzarolo** [comment]
Landed on [v4.0.0](https://github.com/react-native-community/react-native-modal/releases/tag/v4.0.0).  

@iRoachie, the only difference between this version and the previous one in the typedefinition is that the `animationIn` and `animationOut` props now can either be `string` or `{ from: Object, to: Object }` (don't know how to declare it in TS)

**Comment by iRoachie** [comment]
@mmazzarolo Here ya go https://github.com/react-native-community/react-native-modal/pull/73
